### PR TITLE
feat: Add custom prometheus metrics

### PR DESCRIPTION
**Pull request by zarazan**
- Add metric gauge recording of miss_counters, medians, and median deviations at the end of each end_blocker run.
- Add metric gauge recording of exchange_rates each time they are set
- Check for Telemetry.Enabled app.toml setting before recording any end_block metrics

Since voting periods are shorter than 60 seconds we can record each exchange_rate as it is saved to the KVStore. For the miss counters, median, and median deviation metrics they would be pruned before new ones are recorded so they need to be recorded each block.

### CODE REVIEW

File: x/oracle/keeper/end_blocker.go [hunk 1]
```diff
@@ -30,3 +31,26 @@ func (k *Keeper) PruneAllPrices(ctx sdk.Context) {
 func (k *Keeper) IsPeriodLastBlock(ctx sdk.Context, blocksPerPeriod uint64) bool {
 	return (uint64(ctx.BlockHeight())+1)%blocksPerPeriod == 0
 }
+
+func (k *Keeper) RecordEndBlockMetrics(ctx sdk.Context) {
```
Comment (adamewozniak) [inline x/oracle/keeper/end_blocker.go:35]: Doc pls

File: x/oracle/types/price_stamp.go [hunk 2]
```diff
@@ -59,3 +59,18 @@ func (p *PriceStamps) Sort() *PriceStamps {
 	)
 	return &priceStamps
 }
+
+func (p *PriceStamps) NewestPrices() *PriceStamps {
+	blockNum := p.NewestBlockNum()
+	return p.FilterByBlock(blockNum)
+}
+
+func (p *PriceStamps) NewestBlockNum() uint64 {
```
Comment (adamewozniak) [inline x/oracle/types/price_stamp.go:68]: godocs pls

File: util/metrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+func RecordMissCounter(operator sdk.ValAddress, missCounter uint64) {
+	metrics.SetGaugeWithLabels(
+		[]string{"miss_counter"},
+		float32(missCounter),
+		[]metrics.Label{{Name: "address", Value: operator.String()}},
+	)
+}
+
+func RecordExchangeRate(denom string, exchangeRate sdk.Dec) {
```
Comment (adamewozniak) [inline util/metrics/metrics.go:16]: godocs pls

also, let's move some of these metrics labels to consts

File: app/app.go [hunk 4]
```diff
@@ -400,6 +400,7 @@ func New(
 		app.DistrKeeper,
 		app.StakingKeeper,
 		distrtypes.ModuleName,
+		cast.ToBool(appOpts.Get("telemetry.enabled")),
```
Comment (adamewozniak) [inline app/app.go:403]: is there a case where `appOpts.Get` could panic here?

**Review by adamewozniak — APPROVED** [review]
lgtm! just some nitpicks

File: app/app.go (hunk 4 above)
Comment (zarazan) [reply app/app.go:403]: I think it's good, even if telemetry.enabled doesn't exists it doesn't panic. The `cast.ToBool` actually swallows any errors and returns false. This is the suggestions in the cosmos sdk comments "Note, casting Get calls may not yield
	// the expected types and could result in type assertion errors. It is recommend
	// to either use the cast package or perform manual conversion for safety."

**Pull request merged**